### DISCUSSION TITLE
Thermostat Sonoff THR316D SI7021

### DISCUSSION THREAD

**Discussion by salomanov** [arendst/Tasmota#17676] [Q&A]
Hello, please excuse me. I do not know English, I am writing with a translator. I am newbie. I assembled the firmware with the thermostat function. the comment says specify the name of the sensor: THERMOSTAT_SENSOR_NAME "DS18B20" // Name of the local sensor to be used.
i tried also "DS18B20", "SI7021" "THS01"
The thermostat always shows the current temperature at 0. Point out my mistake.

**Comment by barbudor** [comment]
Hello
The name should be the same as the one you see in the SENSOR message

Your SENSOR message should include `  ... "SI7021": { "Temperature": xxx.xx ...} ...` 
So it should be `"SI7021"`

**Reply by sfromis** [reply]
For operational temperature sensor, you need to select the auto-configuration to match your temperature sensor probe. It may be SI7021 (THS01), or DS18B20 depending on what probe you have.

**Comment by salomanov** [comment]
00:00:00.002 HDW: ESP32-D0WD-V3 
00:00:00.074 UFS: FlashFS mounted with 304 kB free
00:00:00.122 CFG: Loaded from File, Подсчет 11
00:00:00.132 QPC: Count 1
00:00:00.339 BRY: Berry initialized, RAM used=4084 bytes
00:00:00.412 Проект Termostat - Tasmota Версия 12.3.1(tasmota)-2_0_5_3(2023-01-11T02:06:45)
00:00:00.418 RSL: RESULT = {"POWER1":"OFF"}
00:00:00.419 RSL: POWER1 = OFF
00:00:00.496 CFG: XDRV87 loaded from file
00:00:00.497 DSP: TM1621
00:00:01.381 WIF: Подключение к AP1 Salomanov Channel 1 BSSId E8:37:7A:91:44:6C в режиме 11n как Termostat-61CDD0-3536...
00:00:03.186 WIF: Соединен
00:00:03.443 HTP: Веб-сервер активен Termostat-61CDD0-3536 с IP-адресом 192.168.1.105
00:00:03.448 RSL: INFO1 = {"Info1":{"Module":"Sonoff THR316D","Version":"12.3.1(tasmota)","FallbackTopic":"cmnd/DVES_61CDD0_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:03.462 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"Termostat-61CDD0-3536","IPAddress":"192.168.1.105"}}
00:00:03.475 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":4}}
00:00:03.478 RSL: RESULT = {"POWER1":"OFF"}
00:00:03.490 RSL: POWER1 = OFF
00:00:03.491 RSL: RESULT = {"POWER2":"OFF"}
00:00:03.492 RSL: POWER2 = OFF
02:52:15.877 QPC: Reset
02:52:17.121 RSL: STATE = {"Time":"2023-01-11T02:52:17","Uptime":"0T00:00:09","UptimeSec":9,"Heap":143,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":45},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Salomanov","BSSId":"E8:37:7A:91:44:6C","Channel":1,"Mode":"11n","RSSI":100,"Signal":-40,"LinkCount":1,"Downtime":"0T00:00:04"}}
02:52:17.161 RSL: SENSOR = {"Time":"2023-01-11T02:52:17","SI7021":{"Temperature":27.9,"Humidity":35.0,"DewPoint":11.0},"Thermostat0":{"ThermostatModeSet":0,"TempTargetSet":18.00,"CtrDutyCycleRead":0},"TempUnit":"C"}
02:52:50.391 CMD: ThermostatModeSet
02:52:50.396 RSL: RESULT = {"ThermostatModeSet1":0}
02:52:53.895 CMD: ThermostatModeSet 1
02:52:53.900 RSL: RESULT = {"ThermostatModeSet1":1}
02:53:07.860 RSL: RESULT = {"POWER1":"ON"}
02:53:07.862 RSL: POWER1 = ON

**Comment by salomanov** [comment]
![2023-01-11_02-53-38](https://user-images.githubusercontent.com/99463178/211670613-2e506dc5-c920-4d3e-90e9-9fa7d1b6881c.png)
![2023-01-11_02-54-37](https://user-images.githubusercontent.com/99463178/211670617-0daa335e-2664-4b19-9f45-c55c45e7995d.png)
![2023-01-11_02-54-55](https://user-images.githubusercontent.com/99463178/211670620-57ee0a42-7927-4c0d-9b97-9e48b710b47a.png)

**Reply by sfromis** [reply]
Ok when you do have the sensor operational as in having the value on the front page, that part of the config should be ok. I read your first post as in this not being the case, but better now.

Your sensor payload also looks good now:
`SENSOR = {"Time":"2023-01-11T02:52:17","SI7021":{"Temperature":27.9,"Humidity":35.0,"DewPoint":11.0},"Thermostat0":{"ThermostatModeSet":0,"TempTargetSet":18.00,"CtrDutyCycleRead":0},"TempUnit":"C"}`

**Comment by salomanov** [comment]
// -- Thermostat control ----------------------------
#define USE_THERMOSTAT                         // Add support for Thermostat
#define THERMOSTAT_CONTROLLER_OUTPUTS 1        // Number of outputs to be controlled independently
#define THERMOSTAT_SENSOR_NAME    "SI7021"        // Name of the local sensor to be used
#define THERMOSTAT_SENSOR_NUMBER 1             // Number of local sensors to be used
#define THERMOSTAT_RELAY_NUMBER 1              // Default output relay number for the first controller (+i for following ones)
#define THERMOSTAT_SWITCH_NUMBER 1             // Default input switch number for the first controller (+i for following ones)

**Comment by salomanov** [comment]
![IMG_20230111_090026.png](https://user-images.githubusercontent.com/99463178/211714898-9d0ab112-8dda-4966-999a-5d2dd8c4bce3.png)

**Comment by salomanov** [comment]
How to make the thermostat take temperature data from si7021

**Reply by barbudor** [reply]
2 possible methods

1) When compiling  your own binary, set  `#define THERMOSTAT_SENSOR_NAME "SI7021"` 

2) Or use a rule:
```
SensorInputSet 0
Rule1 on tele-SI7021#Temperature do TempMeasuredSet %value% endon
Rule1 1
```

**Reply by salomanov** [reply]
point 2 works

**Comment by msjfb** [comment]
I solved the same problem for a DS18B20, I suggest this way for you:
- compile a version with "SI7021" as sensor name  (seems to be already done)
- after a restart, you must send the following 2 commands (in this order) from the console:
    cmnd/YourTasmotaName/ThermostatModeSet 1
    cmnd/YourTasmotaName/SensorInputSet 1

You need to wait 3-4 minutes for the current temperature to update, but it does change and show the current temperature ( at least for me)
It is not instantaneous.

**Reply by salomanov** [reply]
This method doesn't work

**Comment by salomanov** [comment]
another problem appeared, after entering the ThermostatModeSet 1 command, the value from the sensor periodically arrives at 09:18:11.324 RSL: SENSOR = {"Time":"2023-01-14T09:18:11","SI7021":{"Temperature": null,"Humidity":null,"DewPoint":null},"Thermostat0":{"ThermostatModeSet":1,"TempTargetSet":18.00,"CtrDutyCycleRead":66},"TempUnit":"C"}
consequently, the Rule1 command periodically shows a temperature of 0, and the thermostat turns on the relay when it should not

**Reply by barbudor** [reply]
Bad sensor or bad cabling

**Reply by salomanov** [reply]
if ThermostatModeSet": 0, there is no such

**Reply by sfromis** [reply]
As long as you do not have the sensor as such working, the thermostat feature cannot work. This is a surefire indication of the sensor probe not being operational, indeed typically issues with cabling/connection, if the sensor is not broken.
`"SI7021":{"Temperature": null,"Humidity":null,"DewPoint":null}`

Only start checking into the thermostat feature when the sensor is operational.

**Reply by salomanov** [reply]
The fact that the thermostat does not work if there is no indication is clear to me. I am writing about something else. If you do not turn on the thermostat, then temperature data always comes to the console, it is also displayed in the WebUI, but as soon as I turn on the thermostat, it starts periodically NULL, but not for a long time, then it shows and periodically it disappears, then it normally shows. I understand correctly if the contact was bad or the sensor is not working, then without the thermostat function I would observe the same effect that the data would be periodically null, but this is not

**Reply by sfromis** [reply]
What exactly happens in your setup, I do not know, but with electrics and noise being far from simple, it is possible that having the heating on generates noise throwing off the sensor.

You need to get to a situation without `"SI7021":{"Temperature": null,"Humidity":null,"DewPoint":null}` to expect the thermostat feature to work.

**Reply by salomanov** [reply]
The device does not control anything yet, I'm testing it for now

**Reply by salomanov** [reply]
> What exactly happens in your setup, I do not know, but with electrics and noise being far from simple, it is possible that having the heating on generates noise throwing off the sensor.
> 
> You need to get to a situation without "SI7021":{"Temperature": null,"Humidity":null,"DewPoint":null} to expect the thermostat feature to work.

09:13:11.291 RSL: STATE = {"Time":"2023-01-14T09:13:11","Uptime":"0T10:10:34","UptimeSec":36634,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":45},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Salomanov","BSSId":"E8:37:7A:91:44:6C","Channel":1,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":2,"Downtime":"0T00:00:43"}}
09:13:11.302 RSL: SENSOR = {"Time":"2023-01-14T09:13:11","SI7021":{"Temperature":26.9,"Humidity":29.9,"DewPoint":7.8},"Thermostat0":{"ThermostatModeSet":1,"TempTargetSet":18.00,"CtrDutyCycleRead":66},"TempUnit":"C"}
09:18:11.313 RSL: STATE = {"Time":"2023-01-14T09:18:11","Uptime":"0T10:15:34","UptimeSec":36934,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":45},"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Salomanov","BSSId":"E8:37:7A:91:44:6C","Channel":1,"Mode":"11n","RSSI":100,"Signal":-34,"LinkCount":2,"Downtime":"0T00:00:43"}}
09:18:11.324 RSL: SENSOR = {"Time":"2023-01-14T09:18:11","SI7021":{"Temperature":null,"Humidity":null,"DewPoint":null},"Thermostat0":{"ThermostatModeSet":1,"TempTargetSet":18.00,"CtrDutyCycleRead":66},"TempUnit":"C"}
09:23:02.288 RSL: RESULT = {"POWER1":"ON"}

**Reply by sfromis** [reply]
Well, it is controlling the relay. That is not nothing. While I'd not expect that alone to be a cause, maybe in combination with "some weak spot" with the sensor. Whatever the cause is, good sensor data is needed for anything thermostat.

More generally, I've found that Sonoff setup with hiding SI7021 behind a MCU transforming to one-wire to be "problematic", but that is of course no "diagnosis".

**Comment by salomanov** [comment]
It was not possible to launch directly, it works through the Rule, so that it does not produce 0 when the value is null, I use 
SensorInputSet 0
Rule1 on tele-SI7021#Temperature>0 do TempMeasuredSet %value% endon
Rule1 1
